core/playback/device.go

```diff
@@ -0,0 +1,328 @@
+	if err != nil {
+		log.Error("error setting offset", err)
+		return pd.getStatus(), err
+	}
+
+	pd.TrackLoaded = false
+
+	if wasPlaying {
+		_, err = pd.Start()
+		if err != nil {
+			log.Error("error starting new track after skipping")
+			return pd.getStatus(), err
+		}
+	}
+
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) Add(ids []string) (DeviceStatus, error) {
+	log.Debug("processing Add action")
+
+	items := model.MediaFiles{}
+
+	for _, id := range ids {
+		mf, err := pd.ParentPlaybackServer.GetMediaFile(id)
+		if err != nil {
+			return DeviceStatus{}, err
+		}
+		log.Debug("Found mediafile: " + mf.Path)
+		items = append(items, *mf)
+	}
+	pd.PlaybackQueue.Add(items)
+
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) Clear() (DeviceStatus, error) {
+	log.Debug(fmt.Sprintf("processing Clear action on: %s", pd))
+	pd.PlaybackQueue.Clear()
+	pd.TrackLoaded = false
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) Remove(index int) (DeviceStatus, error) {
+	log.Debug("processing Remove action")
+	// pausing if attempting to remove running track
+	if pd.isPlaying() && pd.PlaybackQueue.Index == index {
+		_, err := pd.Stop()
+		if err != nil {
+			log.Error("error stopping running track")
+			return pd.getStatus(), err
+		}
+	}
+
+	if index > -1 && index < pd.PlaybackQueue.Size() {
+		pd.PlaybackQueue.Remove(index)
+	} else {
+		log.Error("Index to remove out of range: " + fmt.Sprint(index))
+	}
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) Shuffle() (DeviceStatus, error) {
+	log.Debug("processing Shuffle action")
+	if pd.PlaybackQueue.Size() > 1 {
+		pd.PlaybackQueue.Shuffle()
+	}
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) SetGain(gain float32) (DeviceStatus, error) {
+	difference := gain - pd.Gain
+	log.Debug(fmt.Sprintf("processing SetGain action. Actual gain: %f, gain to set: %f, difference: %f", pd.Gain, gain, difference))
+
+	pd.adjustVolume(float64(difference) * 5)
+	pd.Gain = gain
+
+	return pd.getStatus(), nil
+}
+
+func (pd *PlaybackDevice) adjustVolume(value float64) {
+	speaker.Lock()
+	pd.Volume.Volume += value
+	speaker.Unlock()
+}
+
+func (pd *PlaybackDevice) Play() {
+	speaker.Lock()
+	pd.Ctrl.Paused = false
+	speaker.Unlock()
+}
+
+func (pd *PlaybackDevice) Pause() {
+	speaker.Lock()
+	pd.Ctrl.Paused = true
+	speaker.Unlock()
+}
+
+func (pd *PlaybackDevice) isPlaying() bool {
+	return pd.TrackLoaded && !pd.Ctrl.Paused
+}
+
+func (pd *PlaybackDevice) loadTrack(mf model.MediaFile) {
+	contentType := mf.ContentType()
+	log.Debug("loading track", "trackname", mf.Path, "mediatype", contentType)
+
+	var streamer beep.StreamSeekCloser
+	var format beep.Format
+	var err error
+
+	switch contentType {
+	case "audio/mpeg":
+		streamer, format, err = decodeMp3(mf.Path)
+	case "audio/x-wav":
+		streamer, format, err = decodeWAV(mf.Path)
+	case "audio/mp4":
+		streamer, format, err = decodeFLAC(*pd.ParentPlaybackServer.GetCtx(), mf.Path)
```

Comment:
Tying the transcoding to the request is a good idea in principle, but there's a problem here: If users (and people will, if they add whole records) set playlists with the SET-Command, than you are disconnected from the context, since the files are loaded and transcoded when needed. This happens from a background goroutine, when the current track ends and the playback continues with the next track ( if there is any ).

One option could be to store the LAST context, which modified the playlist and use that context for the (potential) transcoding ...

Reply:
You are totally right, forget what I said :). The transcoding should NOT be tied to the request at all!